testnet/src/node.rs

```diff
@@ -60,52 +75,149 @@ pub struct Node {
+
+                            let (stacks_header, _) = 
+                                Node::inner_process_tenure(&mined_block, &burn_header_hash, &parent_burn_header_hash,
+                                                           vec![], // no microblocks for now...
+                                                           &mut burndb, &mut chainstate, &dispatcher_channel);
+
+                            let blocks_available = Relayer::load_blocks_available_data(&mut burndb, vec![stacks_header.burn_header_hash])
+                                .expect("Failed to obtain block information for a block we mined.");
+                            if let Err(e) = relayer.advertize_blocks(blocks_available) {
+                                warn!("Failed to advertise new block: {}", e);
+                            }
+                        } else {
+                            warn!("Did not win sortition, my blocks [burn_hash= {}, block_hash= {}], their blocks [par_burn_hash= {}, burn_hash= {}, block_hash ={}]",
+                                  mined_burn_hh, mined_block.block_hash(), parent_burn_header_hash, burn_header_hash, block_header_hash);
+                        }
+                    }
                 },
+                RelayerDirective::RunTenure(registered_key, last_sortitioned_block) => {
+                    last_mined_block = Node::relayer_run_tenure(registered_key, &mut chainstate, last_sortitioned_block,
+                                                                &mut keychain, &mut mem_pool, burn_fee_cap, &mut bitcoin_controller)
+                },
+            }
+        }
+    });
+
+    Ok(())
+}
+
+fn spawn_dispatcher(blocks_path: String, mut event_dispatcher: EventDispatcher,
+                    dispatcher_channel: Receiver<(StacksHeaderInfo, Vec<StacksTransactionReceipt>)>) {
+    thread::spawn(move || {
+        while let Ok((metadata, receipts)) = dispatcher_channel.recv() {
+            let block = {
+                let block_path = StacksChainState::get_block_path(
+                    &blocks_path, 
+                    &metadata.burn_header_hash, 
+                    &metadata.anchored_header.block_hash()).unwrap();
+                StacksChainState::consensus_load(&block_path).unwrap()
             };
 
-            let mut mem_pool = match MemPoolDB::open(
-                false, TESTNET_CHAIN_ID, &stacks_chainstate_path) {
-                Ok(x) => x,
-                Err(e) => {
-                    warn!("Error while connecting to mempool db in peer loop: {}", e);
-                    thread::sleep(time::Duration::from_secs(1));
-                    continue;
-                }
+            let chain_tip = ChainTip {
+                metadata,
+                block,
+                receipts
             };
 
-            this.run(&mut burndb, &mut chainstate, &mut mem_pool, None, poll_timeout)
-                .unwrap();
+            event_dispatcher.process_chain_tip(&chain_tip);
```

Comment:
Why do we choose to have the event_dispatcher in a separate thread, vs relayer thread? Performance?

Reply:
Yeah, I'm confused by this too.  The dispatcher should be passed data synchronously with block processing.

Reply:
As of `67a21e304` -- the dispatcher is passed data synchronously with block processing.